README.md
[![Build Status](https://api.travis-ci.org/kolypto/py-smsframework-clickatell.png?branch=master)](https://travis-ci.org/kolypto/py-smsframework-clickatell)

SMSframework Clickatell Provider
================================

[Clickatell](https://www.clickatell.com/) Provider for [smsframework](https://pypi.python.org/pypi/smsframework/).

You need a "Developers' Central" Clickatell account with an HTTP API set up.
From the API, you need: api_id, username, password.






Installation
============

Install from pypi:

    $ pip install smsframework_clickatell

To receive SMS messages, you need to ensure that
[Flask microframework](http://flask.pocoo.org) is also installed:


    $ pip install smsframework_clickatell[receiver]






Initialization
==============

```python
from smsframework import Gateway
from smsframework_clickatell import ClickatellProvider

gateway = Gateway()
gateway.add_prodiver('clickatell', ClickatellProvider,
    api_id=1,
    user='kolypto',
    pass='123',
    https=False
)
```

Config
------

Source: /smsframework_clickatell/provider.py

* `api_id: str`: API ID to use
* `user: str`: Account username
* `password: str`: Account password
* `https: bool`: Use HTTPS for outgoing messages? Default: `False`






Sending Parameters
==================

Provider-specific sending params:

* `deliv_time: int`: Delay the delivery for X minutes

Example:

```python
from smsframework import OutgoingMessage

gateway.send(OutgoingMessage('+123', 'hi').params(deliv_time=15))
```






Additional Information
======================

OutgoingMessage.meta
--------------------
None.

IncomingMessage.meta
--------------------
* `api_id: str`: API id
* `charset: str`: Message character set (when applicable, else - None)
* `udh: str`: Header Data (when applicable, else - None)

MessageStatus.meta
------------------
* `status: int`: Message status code
* `api_id: str`: API id
* `charge: float`: Charged funds






Public API
==========

ClickatellProvider.get_balance()
--------------------------------
Returns the credist left on the account:

```python
provider = gateway.get_provider('clickatell')
provider.get_balance() #-> 10.6
```






Receivers
=========

Source: /smsframework_clickatell/receiver.py

Message Receiver: /im
---------------------
After a number is purchased, go to Receive Messages > Manage long numbers / short codes, and then click the ‘Edit’
link of the two-way number which you would like to configure. Set "Reply Path" to "HTTP Get" | "HTTP Post",
in the field - put the message receiver URL.

* "Username & Password" is not supported
* "Secondary callback" is up to you

Message Receiver URL: `<provider-name>/im`

Status Receiver: /status
------------------------
To start getting status reports from Clickatell, edit the  HTTP API in the admin panel and click on
"Enable your app to receive message delivery notifications".
In the field, put the receiver URL.

* Status receiver only supports "HTTP Get" and "HTTP Post" methods.
* "basic HTTP Authentication" is not supported

Status Receiver URL: `<provider-name>/status`



CHANGES.md
Changelog
=========

v0.0.1, 2014.01.28
------------------

Initial release


tests/clickatell_test.py
# -*- coding: utf-8 -*-

import unittest
from datetime import datetime

from flask import Flask

from smsframework import Gateway, OutgoingMessage
from smsframework.providers import NullProvider
from smsframework_clickatell import ClickatellProvider

from smsframework_clickatell import error, status


class ClickatellProviderTest(unittest.TestCase):
    def setUp(self):
        # Gateway
        gw = self.gw = Gateway()
        gw.add_provider('null', NullProvider)  # provocation
        gw.add_provider('main', ClickatellProvider, api_id=10, user='kolypto', password='1234')

        # Flask
        app = self.app = Flask(__name__)

        # Register receivers
        gw.receiver_blueprints_register(app, prefix='/a/b/')

    def _mock_response(self, response):
        """ Monkey-patch ClickatellHttpApi so it returns a predefined response """
        def _api_request(method, **params):
            return response
        self.gw.get_provider('main').api._api_request = _api_request

    def test_blueprints(self):
        """ Test blueprints """
        self.assertEqual(
            self.gw.receiver_blueprints().keys(),
            ['main']
        )

    def test_api_request(self):
        """ Test raw requests """
        provider = self.gw.get_provider('main')

        # OK
        self._mock_response('blah-blah')
        self.assertEqual(provider.api_request('hey'), 'blah-blah')

        # Error reported
        self._mock_response('ERR: 1, Auth fail')
        self.assertRaises(error.E001, provider.api_request, 'hey')

        # Error reported, unknown code
        self._mock_response('ERR: 999, Auth fail')
        self.assertRaises(error.ClickatellProviderError, provider.api_request, 'hey')


    def test_getbalance(self):
        """ Test getting balance """
        provider = self.gw.get_provider('main')

        # OK
        self._mock_response('Credit: 12.50')
        self.assertEqual(provider.getbalance(), 12.50)

        # Invalid response
        self._mock_response('blah-blah')
        self.assertRaises(AssertionError, provider.getbalance)

    def test_send(self):
        """ Test message send """
        gw = self.gw

        # OK
        self._mock_response('ID: 11111111')
        message = gw.send(OutgoingMessage('+123456', 'hey', provider='main'))
        self.assertEqual(message.msgid, '11111111')

        # Failure
        self._mock_response('ERR: 001, Auth fail')
        self.assertRaises(error.E001, gw.send, OutgoingMessage('+123456', 'hey', provider='main'))

    def test_receive_message(self):
        """ Test message receipt """

        # Message receiver
        messages = []
        def receiver(message):
            messages.append(message)
        self.gw.onReceive += receiver

        with self.app.test_client() as c:
            # Message 1: artificial
            res = c.get('/a/b/main/im'
                        '?api_id=100'
                        '&moMsgId=1'
                        '&from=123'
                        '&to=456'
                        '&timestamp=2008-08-06 09:43:50'
                        '&charset=ISO-8859-1'
                        '&text=hello there'
                        '&udh=')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(messages), 1)
            message = messages.pop()
            self.assertEqual(message.provider, 'main')
            self.assertEqual(message.msgid, '1')
            self.assertEqual(message.src, '123')
            self.assertEqual(message.dst, '456')
            self.assertEqual(message.body, 'hello there')
            self.assertEqual(message.rtime.strftime('%Y-%m-%d %H:%M:%S'), '2008-08-06 07:43:50')  # UTC
            self.assertEqual(message.meta, {'api_id': '100', 'charset': 'ISO-8859-1', 'udh': ''})

            # Message 2: real, non-unicode
            res = c.get('/a/b/main/im'
                        '?api_id=3460000'
                        '&from=380660000000'
                        '&to=491700000000'
                        '&timestamp=2014-01-29+02%3A08%3A30'
                        '&text=Hi%2C+man'
                        '&charset=ISO-8859-1'
                        '&udh='
                        '&moMsgId=c6b1e0eb9d6b8d549621235aaf089a26')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(messages), 1)
            message = messages.pop()
            self.assertEqual(message.provider, 'main')
            self.assertEqual(message.msgid, 'c6b1e0eb9d6b8d549621235aaf089a26')
            self.assertEqual(message.src, '380660000000')
            self.assertEqual(message.dst, '491700000000')
            self.assertEqual(message.body, 'Hi, man')
            self.assertEqual(message.rtime.strftime('%Y-%m-%d %H:%M:%S'), '2014-01-29 00:08:30')  # UTC
            self.assertEqual(message.meta, {'api_id': '3460000', 'charset': 'ISO-8859-1', 'udh': ''})

            # Message 3: real, unicode
            res = c.get('/a/b/main/im'
                        '?api_id=3460000'
                        '&from=380660000000'
                        '&to=491700000000'
                        '&timestamp=2014-01-29+02%3A05%3A46'
                        '&text=%04%1F%04%40%048%042%045%04B%00%2C%00+%046%04%3E%04%3F%040%00!'
                        '&charset=UTF-16BE'
                        '&udh='
                        '&moMsgId=1de97e5e15f76bb1e374f1ea3d33bf65')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(messages), 1)
            message = messages.pop()
            self.assertEqual(message.provider, 'main')
            self.assertEqual(message.msgid, '1de97e5e15f76bb1e374f1ea3d33bf65')
            self.assertEqual(message.src, '380660000000')
            self.assertEqual(message.dst, '491700000000')
            self.assertEqual(message.body, u'Привет, жопа!')
            self.assertEqual(message.rtime.strftime('%Y-%m-%d %H:%M:%S'), '2014-01-29 00:05:46')  # UTC
            self.assertEqual(message.meta, {'api_id': '3460000', 'charset': 'UTF-16BE', 'udh': ''})

    def test_receive_status(self):
        """ Test status receipt """

        # Status receiver
        statuses = []

        def receiver(status):
            statuses.append(status)

        self.gw.onStatus += receiver

        with self.app.test_client() as c:
            # Status 1: artificial, accepted
            res = c.get('/a/b/main/status'
                        '?from=123'
                        '&to=456'
                        '&status=2'
                        '&api_id=100'
                        '&moMsgId=1'
                        '&charge=0.32')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(statuses), 1)
            st = statuses.pop()
            self.assertEqual(st.msgid, '1')
            self.assertEqual(st.rtime.strftime('%Y-%m-%d'), datetime.utcnow().strftime('%Y-%m-%d'))  # this test fails at midnight :)
            self.assertEqual(st.meta, {'status': 2, 'api_id': '100', 'charge': 0.32})
            self.assertEqual(st.provider, 'main')
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, False)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, 2)
            self.assertEqual(st.status, status.S002.status)
            self.assertEqual(st.error, False)

            # Status 2: artificial, delivered
            res = c.get('/a/b/main/status'
                        '?from=123'
                        '&to=456'
                        '&status=4'
                        '&api_id=100'
                        '&moMsgId=1'
                        '&charge=0.32')
            st = statuses.pop()
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, True)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, 4)
            self.assertEqual(st.status, status.S004.status)
            self.assertEqual(st.error, False)

            # Status 2: artificial, error
            res = c.get('/a/b/main/status'
                        '?from=123'
                        '&to=456'
                        '&status=5'
                        '&api_id=100'
                        '&moMsgId=1'
                        '&charge=0.32')
            st = statuses.pop()
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, False)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, 5)
            self.assertEqual(st.status, status.S005.status)
            self.assertEqual(st.error, True)


smsframework_clickatell/api.py
# -*- coding: utf-8 -*-

import urllib
import urllib2
import re
import math
import binascii

from .const import Features


class ClickatellApiError(RuntimeError):
    def __init__(self, code, message):
        self.code = code
        super(ClickatellApiError, self).__init__(message)


class ClickatellHttpApi(object):
    """ Clickatell HTTP API client """

    def __init__(self, api_id, user, password, https=False):
        """ Create an authenticated client

            :param api_id: Authentication: API ID
            :param user: Authentication: username
            :param password: Authentication: password
            :type https: bool
            :param https: Use HTTPS protocol for requests?
        """
        self._auth = dict(
            api_id=api_id,
            user=user,
            password=password
        )
        self._https = https

        #: Provider API endpoint
        self._hostname = 'api.clickatell.com'

    def _api_request(self, method, **params):
        """ Make an API request and return the result

            :rtype: str
        """
        # Prepare the request
        url = '{schema}://{host}/http/{method}'.format(
            schema='https' if self._https else 'http',
            host=self._hostname,
            method=method
        )
        data = {}
        data.update(self._auth)
        data.update(params)
        post = urllib.urlencode(data)

        # Request
        req = urllib2.Request(url, post)
        res = urllib2.urlopen(req)
        return res.read()

    def api_request(self, method, **params):
        """ Make a custom request to Clickatell and get the response object.

            This also handles errors reported by the Clickatell API

            :type method: str
            :param method: Method name to call
            :param params: Method parameters to send
            :raises HTTPError: Http error code
            :raises URLError: Connection failed
            :raises ClickatellApiError: Clickatell error
        """
        response = self._api_request(method, **params)

        # Error?
        m = re.match(r'^ERR: (\d+), (.*)', response)
        if m:
            raise ClickatellApiError(code=int(m.group(1)), message=m.group(2))
        else:
            return response

    def getbalance(self):
        """ Query balance

            :rtype: float
            :returns: the number of credits available on this particular account.
        """
        response = self.api_request('getbalance')
        m = re.match(r'^Credit: ([\d\.]+)$', response)
        assert m is not None, 'Failed to parse response: {}'.format(response)
        return float(m.group(1))

    def sendmsg(self, to, text, **params):
        """ Send SMS message

            See :meth:`ClickatellHttpApi.api_request` for the list of raised exceptions.

            :param to: Destination number, digits only
            :param text: Message text: str or unicode.
            :param params: Message parameters. See Clickatell docs.

            :param from: Sender address: SenderID, or one of the registered outgoing phone numbers. Default: ''
            :param deliv_ack: Enable/disable delivery acknowledgements. 0|1. Default: 0
            :param callback: Callback URL for status reports. Receives data with HTTP GET. Default: None
            :param deliv_time: SMS delivery delay in minutes. Default: None
            :param escalate: High-pri message: can potentially choose pricier gateways but deliver faster. 0|1. Default: 0
            :param mo: Enable the ability to reply. 0|1. Default: 0.
            :param unicode: Enable unicode messages. If set to 1, the `text` should contain 2-byte unicode. Default: 0
            :param validity: Message validity (expire) period in minutes. Default: None
            :param req_feat: Required features list for the gateway: or'ed constants. see :class:`Features`

            :rtype: str
            :returns: Message id
        """
        # Recipient
        # TODO: Clickatell allows to specify multiple recipients, ','-separated. Implement it! ..and handle the multiple responses
        params['to'] = to

        # Param: `concat`: enable message concatenation.
        # Message length: 160 7bit chars || 140 8bit chars
        # Note: concatenation reduces each message by 7 chars
        concat_parts = (lambda n: 1 if n <= 140 else math.ceil(n/float(140-7)))(len(text.encode('utf-8')))
        if concat_parts > 1:
            params['concat'] = 1

        # CHECKME: seems like req_feat requires FEAT_DELIVACK to be set for acknowledgements. Check it!

        # Unicode message
        if len(text.encode('utf-8')) == len(text):
            params['text'] = str(text)
        else:
            # Unicode message
            params['unicode'] = 1
            params['text'] = binascii.hexlify(text.encode('UTF-16BE'))  # Convert to UCS-2 HEX (utf-16 big-endian)

        # Send it, parse the response
        response = self.api_request('sendmsg', **params)
        m = re.match(r'^ID: (.*)$', response)
        assert m is not None, 'Failed to parse response: {}'.format(response)
        return m.group(1)


smsframework_clickatell/error.py
""" Clickatell error codes """

from smsframework.exc import *


class ClickatellProviderError(ProviderError):
    """ Base class for Clickatell errors

        The __new__ method provides factory behavior: on construct, it mutates to one of its subclasses.
    """
    code = None
    title = '(UNKNOWN ERROR CODE)'

    def __new__(cls, code, message=''):
        # Pick the appropriate class
        for E in cls.__subclasses__():
            if E.code == code:
                C = E
                break
        else:
            C = cls
        return super(ClickatellProviderError, cls).__new__(C, code, message)

    def __init__(self, code, message=''):
        self.code = code
        super(ClickatellProviderError, self).__init__(
            '#{}: {}: {}'.format(self.code, self.title, message)
        )


class E001(ClickatellProviderError, AuthError):
    code = 1
    title = 'Authentication failed'


class E002(ClickatellProviderError, AuthError):
    code = 2
    title = 'Unknown username or password'


class E003(ClickatellProviderError, RequestError):
    code = 3
    title = 'Session ID expired'


class E005(ClickatellProviderError, RequestError):
    code = 5
    title = 'Missing session ID'


class E007(ClickatellProviderError, AuthError):
    code = 7
    title = 'IP Lockdown violation'


class E101(ClickatellProviderError, RequestError):
    code = 101
    title = 'Invalid or missing parameters'


class E102(ClickatellProviderError, RequestError):
    code = 102
    title = 'Invalid user data header'


class E103(ClickatellProviderError, RequestError):
    code = 103
    title = 'Unknown API message ID'


class E104(ClickatellProviderError, RequestError):
    code = 104
    title = 'Unknown client message ID'


class E105(ClickatellProviderError, RequestError):
    code = 105
    title = 'Invalid destination address'


class E106(ClickatellProviderError, RequestError):
    code = 106
    title = 'Invalid source address'


class E107(ClickatellProviderError, RequestError):
    code = 107
    title = 'Empty message'


class E108(ClickatellProviderError, RequestError):
    code = 108
    title = 'Invalid or missing API ID'


class E109(ClickatellProviderError, RequestError):
    code = 109
    title = 'Missing message ID'


class E113(ClickatellProviderError, RequestError):
    code = 113
    title = 'Maximum message parts exceeded'


class E114(ClickatellProviderError, ServerError):
    code = 114
    title = 'Cannot route message'


class E115(ClickatellProviderError):
    code = 115
    title = 'Message expired'


class E116(ClickatellProviderError, RequestError):
    code = 116
    title = 'Invalid Unicode data'


class E120(ClickatellProviderError, RequestError):
    code = 120
    title = 'Invalid delivery time'


class E121(ClickatellProviderError):
    code = 121
    title = 'Destination mobile number blocked'


class E122(ClickatellProviderError):
    code = 122
    title = 'Destination mobile opted out'


class E123(ClickatellProviderError, RequestError):
    code = 123
    title = 'Invalid Sender ID'


class E128(ClickatellProviderError):
    code = 128
    title = 'Number delisted'


class E130(ClickatellProviderError, LimitsError):
    code = 130
    title = 'Maximum MT limit exceeded until <UNIX TIME STAMP>'


class E201(ClickatellProviderError, RequestError):
    code = 201
    title = 'Invalid batch ID'


class E202(ClickatellProviderError, RequestError):
    code = 202
    title = 'No batch template'


class E301(ClickatellProviderError, CreditError):
    code = 301
    title = 'No credit left'


class E901(ClickatellProviderError, ServerError):
    code = 901
    title = 'Internal error'


smsframework_clickatell/status.py
""" Message status codes """

from smsframework.data import *


class ClickatellMessageStatus(MessageStatus):
    status_code = None
    status = '(UNKNOWN STATUS CODE)'

    @classmethod
    def from_code(cls, status_code, **kwargs):
        """ Instantiate one of subclasses by code

            :rtype: type
        """
        for C in cls.__subclasses__():
            if C.status_code == status_code:
                return C(**kwargs)
        return cls(**kwargs)


class S001(ClickatellMessageStatus):
   status_code = 1
   status = 'Message unknown'


class S002(ClickatellMessageStatus, MessageAccepted):
   status_code = 2
   status = 'Message queued'


class S003(ClickatellMessageStatus, MessageAccepted):
   status_code = 3
   status = 'Delivered to gateway'


class S004(ClickatellMessageStatus, MessageDelivered):
   status_code = 4
   status = 'Received by recipient'


class S005(ClickatellMessageStatus, MessageError):
   status_code = 5
   status = 'Error with message'


class S006(ClickatellMessageStatus, MessageError):
   status_code = 6
   status = 'User cancelled message delivery'


class S007(ClickatellMessageStatus, MessageError):
   status_code = 7
   status = 'Error delivering message'


class S008(ClickatellMessageStatus, MessageAccepted):
   status_code = 8
   status = 'OK'


class S009(ClickatellMessageStatus, MessageError):
   status_code = 9
   status = 'Routing error'


class S010(ClickatellMessageStatus, MessageExpired):
   status_code = 10
   status = 'Message expired'


class S011(ClickatellMessageStatus, MessageAccepted):
   status_code = 11
   status = 'Message queued for later delivery'


class S012(ClickatellMessageStatus, MessageError):
   status_code = 12
   status = 'Out of credit'


class S014(ClickatellMessageStatus):
   status_code = 14
   status = 'Maximum MT limit exceeded'


smsframework_clickatell/provider.py
from smsframework import IProvider, exc
from . import error
from . import status
from .api import ClickatellHttpApi, ClickatellApiError
from urllib2 import URLError, HTTPError


class ClickatellProvider(IProvider):
    """ Clickatell provider """

    def __init__(self, gateway, name, api_id, user, password, https=False):
        """ Configure Clickatell provider

            :param api_id: API ID to use
            :param user: Account username
            :param password: Account password
            :param https: Use HTTPS for outgoing messages?
        """
        self.api = ClickatellHttpApi(api_id, user, password, https)
        super(ClickatellProvider, self).__init__(gateway, name)

    def send(self, message):
        """ Send a message

            :type message: smsframework.data.OutgoingMessage.OutgoingMessage
            :rtype: OutgoingMessage
            """
        # Parameters
        params = {}
        if message.src:
            params['from'] = message.src
        if message.provider_options.status_report:
            params['deliv_ack'] = 1
        if message.provider_options.escalate:
            params['escalate'] = 1
        if message.provider_options.allow_reply:
            params['mo'] = 1
        if message.provider_options.expires:
            params['validity'] = message.provider_options.expires
        if message.provider_options.senderId:
            params['from'] = message.provider_options.senderId
        params.update(message.provider_params)

        # Send
        try:
            message.msgid = self.api.sendmsg(message.dst, message.body, **params)
            return message
        except HTTPError as e:
            raise exc.MessageSendError(e.message)
        except URLError as e:
            raise exc.ConnectionError(e.message)
        except ClickatellApiError as e:
            raise error.ClickatellProviderError(e.code, e.message)  # will mutate into the necessary error object

    def make_receiver_blueprint(self):
        """ Create the receiver blueprint

            We do it in a function as the SmsFramework user might not want receivers, consequently, has no reasons
            for installing Flask
        """
        from . import receiver
        return receiver.bp

    #region Public

    def api_request(self, method, **params):
        """ Raw request to Clickatell API

            :rtype: str
            :raises ConnectionError: Connection error
            :raises MessageSendError: HTTP error
            :raises ClickatellProviderError: Error with the request
        """
        try:
            return self.api.api_request(method, **params)
        except HTTPError as e:
            raise exc.MessageSendError(e.message)
        except URLError as e:
            raise exc.ConnectionError(e.message)
        except ClickatellApiError as e:
            raise error.ClickatellProviderError(e.code, e.message)  # will mutate into the necessary error object

    def getbalance(self):
        """ Query balance

            :rtype: float
            :returns: The number of credits available
        """
        try:
            return self.api.getbalance()
        except HTTPError as e:
            raise exc.MessageSendError(e.message)
        except URLError as e:
            raise exc.ConnectionError(e.message)
        except ClickatellApiError as e:
            raise error.ClickatellProviderError(e.code, e.message)

    #endregion


smsframework_clickatell/receiver.py
from datetime import datetime, timedelta

from flask import Blueprint
from flask.globals import request, g

from smsframework.data import IncomingMessage
from .status import ClickatellMessageStatus

bp = Blueprint('smsframework-clickatell', __name__, url_prefix='/')


def _merge_request(request):
    """ Merge query string and form body """
    data = {}
    data.update(request.form.to_dict())
    data.update(request.args.to_dict())
    return data


@bp.route('/im')
def im():
    """ Incoming message handler

        Clickatell sends data with either GET or POST:

        * api_id: Api ID
        * moMsgId: MO message ID
        * from: Originating ISDN
        * to: Destination ISDN
        * timestamp: Date & Time in MySQL format, GMT+0200: "2008-08-06 09:43:50"
        * charset: DCS Character Coding [when applicable]
        * udh: Header Data [e.g. UDH etc.] [when applicable]
        * text: Message Data
    """
    req = _merge_request(request)

    # Check fields
    for n in ('api_id', 'moMsgId', 'from', 'to', 'timestamp', 'charset', 'udh', 'text'):
        assert n in req, 'Clickatell sent a message with missing "{}" field: {}'.format(n, req)

    # Parse date
    rtime = datetime.strptime(req['timestamp'], '%Y-%m-%d %H:%M:%S')
    rtime -= timedelta(hours=2)  # Date is in GMT+0200. Alter it to UTC

    # Message encoding
    req['text'] = req['text'].decode(req['charset'])

    # IncomingMessage
    message = IncomingMessage(
        src=req['from'],
        body=req['text'],  # CHECKME: test that unicode works
        msgid=req['moMsgId'],
        dst=req['to'],
        rtime=rtime,
        meta={
            'api_id': req['api_id'],
            'charset': req['charset'],
            'udh': req['udh']
        }
    )

    # Process it
    provider = g.provider  # yes, this is how the current provider is fetched
    " :type: smsframework.IProvider.IProvider "
    provider._receive_message(message)  # any exceptions will respond with 500, and Clickatell will happily retry later

    # Ack
    return 'OK'  # Clickatell protocol is well-structured, yes


@bp.route('/status')
def status():
    """ Incoming status report

         Clickatell sends data with either GET or POST:

         * from: source number
         * to: destination number
         * status: status code
         * cliMsgId: client-specified msgid (if provided)
         * api_id: API id
         * moMsgId: msgid
         * charge: charged credits
    """
    req = _merge_request(request)

    # Check fields
    for n in ('from', 'to', 'status', 'api_id', 'moMsgId', 'charge'):
        assert n in req, 'Clickatell sent a status with missing "{}" field: {}'.format(n, req)

    # MessageStatus
    status = ClickatellMessageStatus.from_code(
        int(req['status']),
        msgid=req['moMsgId'],
        meta={
            'status': int(req['status']),
            'api_id': req['api_id'],
            'charge': float(req['charge'])
        }
    )

    # Process it
    g.provider._receive_status(status)  # exception respond with http 500

    # Ack
    return 'OK'
